File: 1544-count-good-nodes-in-binary-tree/count-good-nodes-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int count = 0;
    void dfs(TreeNode* root, int max_val)
    {
        if(!root) return;
        int val = root->val;
        if(val >= max_val)
        {
            count+=1;
            max_val = max(val, max_val);
        }
        if(root->left)
        {
            dfs(root->left, max_val);
        }
        if(root->right)
        {
            dfs(root->right, max_val);
        }
    }
    int goodNodes(TreeNode* root) {
        dfs(root, root->val);
        return count;
    }
};

File: 1544-count-good-nodes-in-binary-tree/count-good-nodes-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    count = 0
    
    def dfs(self, root, curr_val):
        if not root:
            return
        if root.val >= curr_val:
            self.count += 1
            curr_val = max(curr_val, root.val)
        self.dfs(root.left, curr_val)
        self.dfs(root.right, curr_val)
        
    def goodNodes(self, root: TreeNode) -> int:
        self.dfs(root, float('-inf'))
        return self.count
